File: 2024/Day_09/part02.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <algorithm>
using namespace std;

vector<int> read_file() {
    ifstream file("test.txt");
    if (!file.is_open()) {
        cerr << "Error Opening File" << endl;
        exit(-1);
    }

    string line;
    vector<int> data;
    while(getline(file, line)) {
        for (char ch : line)
            data.push_back(ch - '0');  // char to int
    }

    if (data.size() % 2 != 0) data.push_back(0);
    file.close();
    return data;
}

vector<pair<int, int>> make_pairs(vector<int> &data) {
    vector<pair<int, int>> pairedFiles;

    for (size_t i = 0; i < data.size(); i += 2) {
        pairedFiles.push_back({data[i], data[i + 1]});
    }

    return pairedFiles;
}

void harddrive_map(const vector<pair<int, int>> &pairedFiles, vector<int> &harddrive, unordered_map<int, int> &file_dict, int &largest_id) {
    int idVal = 0;

    for (int i = 0; i < pairedFiles.size(); i++) {
        harddrive.insert(harddrive.end(), pairedFiles[i].first, idVal);
        harddrive.insert(harddrive.end(), pairedFiles[i].second, -1);
        file_dict[idVal] = pairedFiles[i].first;
        idVal++;
    }

    largest_id = idVal - 1;
}

int num_consecutive_space(const vector<int>& harddrive, int itr) {
    int consecutive_count = 0;
    for (size_t i = itr; i < harddrive.size(); i++) {
        if (harddrive[i] != -1) break;
        consecutive_count++;
    }
    return consecutive_count;
}

void optimize_space(vector<int> &harddrive, const unordered_map<int, int> &fileDict, int largestID) {
    int descID = largestID;

    while (descID > 0) {
        size_t i = 0;
        while (i < harddrive.size()) {
            if (harddrive[i] == descID) break;
            if (harddrive[i] == -1) {
                int num_space = num_consecutive_space(harddrive, i);
                if (num_space < fileDict.at(descID)) {
                    i++;
                    continue;
                } else {
                    size_t valLoc = find(harddrive.begin(), harddrive.end(), descID) - harddrive.begin();
                    for (int idx = 0; idx < fileDict.at(descID); idx++)
                        swap(harddrive[i + idx], harddrive[valLoc + idx]);
                    break;
                }
            }
            i++;
        }
        descID--;
    }
}

long long calculate_checksum(vector<int> &optHarddrive) {
    long long checksum = 0;
    for (size_t idx = 0; idx < optHarddrive.size(); ++idx) {
        if (optHarddrive[idx] == -1) checksum += idx * 0;
        else checksum += idx * optHarddrive[idx];
    }
    return checksum;
}



int main() {
    vector<int> data = read_file();
    vector<pair<int, int>> pairedFiles = make_pairs(data);

    vector<int> harddrive;
    unordered_map<int, int> fileDict;
    int largestID;
    harddrive_map(pairedFiles, harddrive, fileDict, largestID);
    optimize_space(harddrive, fileDict, largestID);

    long long checksum = calculate_checksum(harddrive);
    cout << checksum << endl;
    return 0;
}

File: 2024/Day_09/part02.py
import numpy as np

# reads in data from file
with open("test.txt", "r") as file:
    data = np.array(list(file.read()), dtype=int)
    # adds 0 to last file with NO partitions after
    if len(data) % 2 != 0:
        data = np.append(data, 0)

# create partition map of harddrive
def harddrive_map(paired_files):
    harddrive = []
    file_dict = {}
    id_val = 0

    for pair in paired_files:
        harddrive.extend([id_val] * pair[0])
        harddrive.extend([-1] * pair[1])
        file_dict[id_val] = pair[0]
        id_val += 1
        

    return harddrive, file_dict, id_val-1

def num_consecutive_space(harddrive, itr):
    consecutive_count = 0
    for i in range(itr, len(harddrive)):
        if harddrive[i] != -1: break
        consecutive_count += 1 
    return consecutive_count

# shifts far right files to furthest empty left
def optimize_space(harddrive, file_dict, largest_id):
    desc_id = largest_id

    while(desc_id > 0):
        i = 0
        while (i < len(harddrive)):
            if harddrive[i] == desc_id: break
            if harddrive[i] == -1:
                num_space = num_consecutive_space(harddrive, i)
                if num_space < file_dict[desc_id]:
                    i += 1
                    continue
                else:
                    val_loc = harddrive.index(desc_id)
                    for idx in range(file_dict[desc_id]):
                        harddrive[i+idx], harddrive[val_loc+idx] = harddrive[val_loc+idx], harddrive[i+idx]
                    break
            i += 1
        desc_id -= 1


def calculate_checksum(opt_harddrive):
    checksum = 0
    for idx, val in enumerate(opt_harddrive):
        if val == -1: checksum += idx * 0
        else: checksum += idx * val
    return checksum

paired_files = data.reshape(-1, 2)          # pairs 1st and 2nd elements together
harddrive, file_dict, largest_id = harddrive_map(paired_files)
optimize_space(harddrive, file_dict, largest_id)
checksum = calculate_checksum(harddrive)
print(checksum)

